### PR TITLE
 Remove `{,un}register_observer` from `FileLike`

### PR DESCRIPTION
**Pull request by lrh2000**
~~*Depends on https://github.com/asterinas/asterinas/pull/1496 and https://github.com/asterinas/asterinas/pull/1498 to avoid conflicts.*~~

This PR closes https://github.com/asterinas/asterinas/issues/1355:
 - https://github.com/asterinas/asterinas/issues/1355

A brief summary of the most important changes:
 - **Before this PR:**
   - `Poller` is a special kind of `Observer` that waits for certain events. It can be _conditionally_ registered with `FileLike::poll`.
   - Other kinds of `Observer`s must be registered with `FileLike::register_observer`.
 - **After this PR:**
   - `RawPoller` can represent arbitrary kinds of `Observer`. It can be _unconditionally_ registered with `FileLike::poll`.
     - However, `RawPoller` has generic parameters, so it cannot be used in object-safe traits. `AnyPoller` is used to erase the generic argument.
   - `Poller` is a special kind of `RawPoller` that can be used to wait for certain events.

This way, I believe no functionality is lost. However, the `FileLike` APIs are simplified.

API overview:

```rust
pub struct RawPoller<T>;
pub struct AnyPoller;

impl<T: Observer<IoEvents> + 'static> RawPoller<T> {
    pub fn with_observer(observer: T) -> Self;
    pub fn observer(&self) -> &Arc<T>;
    pub fn as_any_mut(&mut self) -> &mut AnyPoller;
}

pub trait Pollable {
    fn poll(&self, mask: IoEvents, poller: Option<&mut AnyPoller>) -> IoEvents;
}
```

```rust
pub struct Poller;

impl Poller {
    pub fn new() -> Self;
    pub fn as_any_mut(&mut self) -> &mut AnyPoller;
    pub fn wait(&self, timeout: Option<&Duration>) -> Result<()>;
}
```

### CODE REVIEW

File: kernel/src/process/signal/poll.rs [hunk 1]
```diff
@@ -37,34 +37,29 @@ impl Pollee {
...
+    /// This operation is _atomic_ in the sense that if there are interesting events, either the
+    /// events are returned or the poller is notified.
+    pub fn poll(&self, mask: IoEvents, poller: Option<&mut AnyPoller>) -> IoEvents {
         let mask = mask | IoEvents::ALWAYS_POLL;
 
-        // Fast path: return events immediately
-        let revents = self.events() & mask;
-        if !revents.is_empty() || poller.is_none() {
-            return revents;
-        }
```
Comment (StevenJiang1110) [inline kernel/src/process/signal/poll.rs:55]: Why the fast path is removed? It seems the fast path makes sense?

File: kernel/src/process/signal/poll.rs [hunk 2]
```diff
@@ -131,75 +126,138 @@ impl Pollee {
...
+    // The inner with observer type erased.
+    inner: AnyPoller,
 }
 
-impl Default for Poller {
-    fn default() -> Self {
-        Self::new()
+impl<T: Observer<IoEvents> + 'static> Poller<T> {
+    /// Constructs a new poller with the specified observer.
+    pub fn with_observer(observer: T) -> Self {
```
Comment (StevenJiang1110) [inline kernel/src/process/signal/poll.rs:141]: have some skepticism about whether users would want to construct a `Poller` instead of a `WaitablePoller`. After searching this PR, I found that `Poller::with_observer` is called only once (and not within `WaitablePoller`) in the `set_owner` method. Additionally, if we use `WaitablePoller`, it seems that there would be no drawbacks. What if we simply expose `WaitablePoller` to users and keep `Poller` as an internal API? In that case, it might make more sense to rename `WaitablePoller` to just `Poller` for the sake of user simplicity.

File: kernel/src/process/signal/poll.rs [hunk 3]
```diff
@@ -131,75 +126,138 @@ impl Pollee {
...
-    /// Waits until some interesting events happen since the last wait or until the timeout
-    /// expires.
+    /// Gets a mutable reference to the type-erased poller.
     ///
-    /// The waiting process can be interrupted by a signal.
-    pub fn wait(&self, timeout: Option<&Duration>) -> Result<()> {
-        self.event_counter.read(&self.waiter, timeout)?;
-        Ok(())
+    /// See [`AnyPoller`] for more details.
+    pub fn as_any(&mut self) -> &mut AnyPoller {
```
Comment (StevenJiang1110) [inline kernel/src/process/signal/poll.rs:158]: Naming as `as_any_mut` may be better. And the same as `WaitablePoller::as_any`

**Review by StevenJiang1110 — COMMENTED** [review]
Good job! Apologies for the late response.

File: kernel/src/process/signal/poll.rs (hunk 1 above)
Comment (lrh2000) [reply kernel/src/process/signal/poll.rs:55]: There are several reasons why I remove this fast path. I'll list its pros and cons here, and we'll have to decide between the tradeoffs.

### Pros
 1. The semantics of the `poll` method is unclear if we keep this fast path.
    - As a concrete example, it is not possible for the caller to know if the poller is registered by checking the return value of the `poll` method.
       - If the return value contains the events of interest, (i) the poller may not be registered when the fast path is used, meanwhile, (ii) the poller may be registered when the fast path is not used and a race condition occurs later.
 2. Because of the first point, it will not be possible for us to remove the `register_observer` and `unregister_observer` methods directly.
     - This is because we need a method to unconditionally register the observer (i.e. `register_observer`), but the specific semantics cannot be expressed by the `poll` method with its fast path.
 3. Because of the first point, we need to call the `poll` method in the loop of the `wait_events` method. (Otherwise the poller cannot be guaranteed to be registered). This introduces some strangeness in two ways:
    - This leads to redundant registrations. We call the `poll` methods repeatedly, so the poller can be registered multiple times.
    - After implementing https://github.com/asterinas/asterinas/issues/1356, a call to the `poll` method becomes slightly more expensive (as it may involve calculating I/O events). Repeatedly calling the `poll' method in the `wait_events` loop is simply not necessary and should be avoided.
 4. If we're going to make the pollee completely stateless (https://github.com/asterinas/asterinas/issues/1356#issue-2537879395), this fast path will simply become impossible. We will no longer maintain the I/O events in the pollee, so the fast path has to be removed.
    - If we make the pollee conditionally stateless (https://github.com/asterinas/asterinas/issues/1356#issuecomment-2436849331), the fast path can be preserved. But the fast path can only take effect if there are cached I/O events, which seems to make the semantics of the `poll` method even more obscure.

### Cons

Removing a fast path can hurt performance. However, for the specific fast path, I think the damage is minimal.
 - For the epoll file, the poller must be registered regardless of whether the I/O events of interest are present or not. So the original fast path makes no sense.
 - For the `wait_events` method, before we call the `poll` method, we already had another fast path that called the operation (`try_op`) directly. So the original fast path makes no sense if there are no race conditions (but race conditions are rare, so they can be ignored).
https://github.com/asterinas/asterinas/blob/1f612e45f7fccb4caa3901ee26605f828ec64a06/kernel/src/process/signal/poll.rs#L266-L270
 - The original fast path only makes sense for the `poll`/`select` system call and the `do_poll` method. In this scenario, if we see some I/O events of interest, we don't want to keep registering the poller, instead we will return the I/O events directly to the user. However, this scenario occurs at most once during a `do_poll` method, because after that we'll provide `None` as the `poller`.
https://github.com/asterinas/asterinas/blob/1f612e45f7fccb4caa3901ee26605f828ec64a06/kernel/src/syscall/poll.rs#L69-L73

### Conclusion

So in conclusion, I think removing the fast path brings several pros and only one con. For the only con, the damage should also be minimal.

File: kernel/src/process/signal/poll.rs (hunk 2 above)
Comment (lrh2000) [reply kernel/src/process/signal/poll.rs:141]: > I found that `Poller::with_observer` is called only once

`Pollers` are designed to be used in the epoll file and in the `SIGIO` delivery. However, in this PR, the epoll file uses `AnyPoller` directly. In principle, it should use `Poller` instead (we may want to discuss this in https://github.com/asterinas/asterinas/pull/1500).

> Additionally, if we use `WaitablePoller`, it seems that there would be no drawbacks.

I do not see how this is possible. A `Poller` can be with any observer, and the observer may not support waiting for events. This is the case for the epoll file and the `SIGIO` package.

> In that case, it might make more sense to rename `WaitablePoller` to just `Poller` for the sake of user simplicity.

Since `WaitablePoller` is used more often, I agree that it deserves a short name. A possible way is to rename:
 - `WaitablePoller` -> `Poller`
 - `Poller` -> `RawPoller`

File: kernel/src/process/signal/poll.rs (hunk 1 above)
Comment (StevenJiang1110) [reply kernel/src/process/signal/poll.rs:55]: Thank you for your explanation!

I understand that you are trying to implement the `poll` method to ensure that the `Poller` is registered, which essentially serves as a superset of the original `register_observer`. The potential performance benefits of the fast path also make sense to me. Therefore, we can remove the fast path in this case.

However, I’m unclear about the points concerning the stateless pollee. We can revisit this topic later, once the design is finalized.

File: kernel/src/process/signal/poll.rs (hunk 2 above)
Comment (StevenJiang1110) [reply kernel/src/process/signal/poll.rs:141]: > I do not see how this is possible. A Poller can be with any observer, and the observer may not support waiting for events. This is the case for the epoll file and the SIGIO delivery.

I misunderstand the meaning of `Poller`. Then your use should be correct.

> Since WaitablePoller is used more often, I agree that it deserves a short name. A possible > > way is to rename:

> - WaitablePoller -> Poller
> - Poller -> RawPoller

Yes, the renaming will be better.

File: kernel/src/process/signal/poll.rs (hunk 1 above)
Comment (lrh2000) [reply kernel/src/process/signal/poll.rs:55]: Also note that we're doing exactly the same thing as Linux (Linux does not have a fast path for its poll implementation, see [`tcp_poll`](https://elixir.bootlin.com/linux/v6.11/source/net/ipv4/tcp.c#L507) for example). So performance shouldn't be an issue as long as we are comparing to Linux).

File: kernel/src/process/signal/poll.rs (hunk 2 above)
Comment (lrh2000) [reply kernel/src/process/signal/poll.rs:141]: Renamed and also updated the PR description.

**Review by tatetian — COMMENTED** [review]
The PR looks good me for the most part. Merging observers into pollers does simplify the code when it is possible to do so.

I am a little bit unsatisfied with the names of `Poller`, `RawPoller<T>`, and `AnyPoller`. To me, a poller must be able to be used with `Pollee` to do poll. This requires it to have a `wait` method. So in my opinion, only `Poller` can be called a poller. RawPoller<T>` and `AnyPoller` are not qualified.

I suggest renaming `RawPoller<T>` to `PollerLikeObserver<O>` and `AnyPoller` to `PollerLike`.

```rust
/// An adaptor to make an observer acceptable as the argument for the `Pollee::poll` method.
pub struct ObserverPollAdaptor<O> { ... }

/// An opaque object used as the argument of the `Pollee::poll` method.
///
/// This "opaque" type unifies two classes of entities that may be interested
/// in the `IoEvents` on a `Pollee` object:
/// 1. `Poller`, and
/// 2. `ObserverPollAdaptor<O: Observer<IoEvents>>`.
///
/// When this object is dropped, the entity will no longer be notified of the events from the `Pollee`.
pub struct PollHandle { ... }
```

Since the type names have been changed, the method `as_any_mut` is no longer appropriate. 

```rust
self.poll(mask, Some(poller.as_any_mut()))
```

I suggest renaming `as_any_mut` to `as_mut_handle`.

```rust
self.poll(mask, Some(poller.as_any_handle()))
```

**Comment by lrh2000** [comment]
> I suggest renaming `RawPoller<T>` to `ObserverPollAdaptor<O>` and `AnyPoller` to `PollHandle`.

I think `ObserverPollAdaptor` and `PollHandle` seem to be accurate names. My only concern is that `ObserverPollAdaptor` seems to be too long. Not sure if we can make it shorter, though.

> ```rust
> /// This "opaque" type unifies two classes of entities that may be interested
> /// in the `IoEvents` on a `Pollee` object:
> /// 1. `Poller`, and
> /// 2. `ObserverPollAdaptor<O: Observer<IoEvents>>`.
> /// Both of them provide a method to present themselves as a `PollHandle`.
> ```

There is nothing special about `Poller`. `Poller` internally just uses `ObserverPollAdaptor` with a special observer that supports threads to wait on. So I'm afraid the claim that `Poller` and `ObserverPollAdaptor` are two classes of entities that can generate a `PollHandle` seems a bit odd (but still acceptable if we only look at the public APIs).

**Comment by tatetian** [comment]
> > I suggest renaming `RawPoller<T>` to `ObserverPollAdaptor<O>` and `AnyPoller` to `PollHandle`.
> 
> I think `ObserverPollAdaptor` and `PollHandle` seem to be accurate names. My only concern is that `ObserverPollAdaptor` seems to be too long. Not sure if we can make it shorter, though.
> 
> > ```rust
> > /// This "opaque" type unifies two classes of entities that may be interested
> > /// in the `IoEvents` on a `Pollee` object:
> > /// 1. `Poller`, and
> > /// 2. `ObserverPollAdaptor<O: Observer<IoEvents>>`.
> > /// Both of them provide a method to present themselves as a `PollHandle`.
> > ```
> 
> There is nothing special about `Poller`. `Poller` internally just uses `ObserverPollAdaptor` with a special observer that supports threads to wait on. So I'm afraid the claim that `Poller` and `ObserverPollAdaptor` are two classes of entities that can generate a `PollHandle` seems a bit odd (but still acceptable if we only look at the public APIs).

Yeah, `ObserverPollAdaptor` is a bit of long. I am not proud of that. The good news is that it is rarely used. So it should be acceptable. Name accuracy triumphs name lengths. That is my preference in naming.

**Comment by lrh2000** [comment]
> Yeah, `ObserverPollAdaptor` is a bit of long. I am not proud of that. The good news is that it is rarely used. So it should be acceptable. Name accuracy triumphs name lengths. That is my preference in naming.

I am going to call it `PollAdaptor`. My reasoning is:
 - The observer is already reflected in the trait bound, so repeating it in the type name is kind of redundant.
 - Let's look at a real use case, i.e., the code that delivers `SIGIO` to the file owner. I would like to see something like `PollAdaptor<OwnerObserver>` _instead of_ `ObserverPollAdaptor<OwnerObserver>` (which is just ugly because the observer appears twice).

Please let me know if you think this causes ambiguity. I can then re-add the word `Observer`.
